src/lib/core/datetime.c

```diff
@@ -67,6 +67,30 @@ datetime_strftime(const struct datetime *date, char *buf, size_t len,
 	return tnt_strftime(buf, len, fmt, &tm);
 }
 
+void
+tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
+{
+	assert(tm->tm_mday > 0 && tm->tm_mday < 32);
+	assert(tm->tm_mon >= 0 && tm->tm_mon <= 11);
+	dt_t dt = dt_from_ymd(tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday);
+	int64_t local_secs =
+		(int64_t)dt * SECS_PER_DAY - SECS_EPOCH_1970_OFFSET;
+	local_secs += tm->tm_hour * 3600 + tm->tm_min * 60 + tm->tm_sec;
+	date->epoch = local_secs - tm->tm_gmtoff;
+	date->nsec = tm->tm_nsec;
+	date->tzindex = 0;
+	date->tzoffset = tm->tm_gmtoff / 60;
+}
+
+char *
+datetime_strptime(struct datetime *date, const char *buf, const char *fmt)
+{
+	struct tnt_tm t = { .tm_epoch = 0 };
+	char * ret = tnt_strptime(buf, fmt, &t);
```

Comment:
Please, handle `ret == NULL` here.

Otherwise a failed parse fires an assertion:
```
tarantool> dt.parse('17001213', {format='%F'})
tarantool: /home/sergey/Source/tarantool/src/lib/core/datetime.c:73: tm_to_datetime: Assertion `tm->tm_mday > 0 && tm->tm_mday < 32' failed.

```

Reply:
Yes, good point!

Reply:
```diff
commit d3366ff5567a13a4dacc684e7b0d45eea402c3a4
Author: Timur Safin <tsafin@tarantool.org>
Date:   Tue Jan 18 00:12:59 2022 +0300

    temp - indenting, assert and proper NULL handling in tnt_strptime

diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 477b742c2..aecac58ab 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -60,8 +60,9 @@ datetime_to_tm(const struct datetime *date, struct tnt_tm *tm)
 
 size_t
 datetime_strftime(const struct datetime *date, char *buf, size_t len,
-		      const char *fmt)
+		  const char *fmt)
 {
+	assert(date != NULL);
 	struct tnt_tm tm;
 	datetime_to_tm(date, &tm);
 	return tnt_strftime(buf, len, fmt, &tm);
@@ -87,7 +88,8 @@ datetime_strptime(struct datetime *date, const char *buf, const char *fmt)
 {
 	struct tnt_tm t = { .tm_epoch = 0 };
 	char * ret = tnt_strptime(buf, fmt, &t);
-	tm_to_datetime(&t, date);
+	if (ret != NULL)
+		tm_to_datetime(&t, date);
 	return ret;
 }
 
```

And thus we have:
```
tarantool> date.parse('17001213', {format='%F'})
---
- error: 'builtin/datetime.lua:929: Could not parse ''17001213'' using ''%F'' format'
...
```